src/signal.cpp

```diff
@@ -264,24 +264,25 @@ void signal_set_handlers() {
     // an event.
     sigaction(SIGINT, &act, 0);
     sigaction(SIGQUIT, &act, 0);
-    sigaction(SIGTSTP, &act, 0);
-    sigaction(SIGTTIN, &act, 0);
-    sigaction(SIGTTOU, &act, 0);
     sigaction(SIGCHLD, &act, 0);
 
     // Ignore sigpipe, which we may get from the universal variable notifier.
     sigaction(SIGPIPE, &act, 0);
 
+    // Ignore TTY I/O signals as we'll handle such problems by appropriate checks of relevant
+    // syscall return values.
+    act.sa_handler = SIG_IGN;
+    sigaction(SIGTSTP, &act, 0);
+    sigaction(SIGTTIN, &act, 0);
+    sigaction(SIGTTOU, &act, 0);
+
```

Comment:
That code in fish seems to mimic bash's `initialize_job_control()` which also does a `kill (0, SIGTTIN)` if the `tcgetpgrp()` fails. The only reason I'm ignoring SIGTTIN is because it it related to SIGTTOU which needs to be ignored to avoid a pathological failure case when fish writes to a broken tty. However, reading from that broken tty should probably not be protected in the same manner. I'll revert that portion of the change.